### DISCUSSION TITLE
Micronaut Framework 4.3.0 Plan

### DISCUSSION THREAD

**Discussion by sdelamo** [micronaut-projects/micronaut-core#10137] [Planning]
We released [4.2.0 last Friday](https://micronaut.io/2023/11/17/micronaut-framework-4-2-0-released/).  

I have scheduled Micronaut Framework 4.3.0 for January 15th, 2024. 

We use [Github Projects](https://github.com/orgs/micronaut-projects/projects) to plan the work. We have a [GitHub Project for 4.3.0](https://github.com/orgs/micronaut-projects/projects/68). 

## 4.3.0 Plan
My recommendation for 4.3.0 is to continue to work on releasing the work in progress. We will continue triaging issues and closing bugs in the different modules during this cycle. 

#### Youtube Channel
Release ten videos of less than 15m for the YouTube channel covering features of the framework. We have to improve our presence on YouTube. Many people only learn with video nowadays. 

#### Guides
Release 4 Guides comparing Micronaut Framework to Spring Boot. It is vital to show how easy it is to move from Spring Boot to Micronaut Framework.

### New Modules

I think we may be able to release three new modules we are currently working - some of these modules may go in the 4.4.0 release: 

- [**Load Shedding**](https://github.com/micronaut-projects/micronaut-core/issues/9293)
- [**Chatbots**](https://micronaut-projects.github.io/micronaut-chatbots/snapshot/guide/)
- [**SourceGen**](https://github.com/micronaut-projects/micronaut-sourcegen/issues)

### Finish in progress Features

We have other features in progress which I would like to finish: 
- [**AWS** Better integration with DynamoDB](https://github.com/micronaut-projects/micronaut-aws/pull/1664). 

#### Expression Language
- [Add conditional caching with the new EL](https://github.com/micronaut-projects/micronaut-cache/issues/572). [Status: in-progress PR #692](https://github.com/micronaut-projects/micronaut-cache/pull/692) 
- [Support Expression Language for Tenant resolution](https://github.com/micronaut-projects/micronaut-multitenancy/issues/248)

#### Build
- [Integration with synk](https://github.com/micronaut-projects/micronaut-core/issues/10100)

#### Eclipse Store
- [Fork MicroStream repository to support EclipseStore](https://github.com/micronaut-projects/micronaut-microstream/issues/299)
#### SQL 
- [Update to Hibernate 6.3.1.final](https://github.com/micronaut-projects/micronaut-sql/pull/1080)

## Backlog

The [GitHub Project "Roadmap Top Level Features"](https://github.com/orgs/micronaut-projects/projects/69) contains high-level features we could consider:

### Micronat Launch

- [Changes to Micronaut Launch](https://github.com/micronaut-projects/micronaut-starter/issues/2191)

### Micronaut Data
The EL improvements mentioned above should help with this feature: 

- [Support Multitenancy via discriminator column](https://github.com/micronaut-projects/micronaut-data/issues/2577)

### Views
- [**in-progress** Scaffolding Feature](https://github.com/micronaut-projects/micronaut-views/issues/606)
- [Integrate with HTMX](https://github.com/micronaut-projects/micronaut-views/issues/660)

### Core
- [Block detection for methods annotated with `@Blocking`	](https://github.com/micronaut-projects/micronaut-core/issues/10020)

### Security
- [Add nonreactive API for authentication fetcher](https://github.com/micronaut-projects/micronaut-security/issues/1476).  
- [**in-progress** Support Paseto tokens](https://github.com/micronaut-projects/micronaut-security/pull/813).

### GraphQL
- [Apollo GraphQL Federation support	](https://github.com/micronaut-projects/micronaut-graphql/issues/398)

### Micronaut CRaC 
- [Consider exposing an endpoint that creates a CRaC checkpoint programmatically when invoked](https://github.com/micronaut-projects/micronaut-crac/issues/116)

### New Modules/Integrations

- [Micronaut Pay](https://github.com/micronaut-projects/micronaut-core/issues/10134)
- [OpenSearch Module](https://github.com/micronaut-projects/micronaut-elasticsearch/issues/443)
- [rSocket support](https://github.com/micronaut-projects/micronaut-core/issues/2002)
- [Integration with Langchain4j](https://github.com/micronaut-projects/micronaut-core/issues/10009)
- [Better Integration with GraalVM Truffle](https://github.com/micronaut-projects/micronaut-core/issues/10012)

Any feedback about what we should consider for 4.3.0 is more than welcome.

**Comment by scprek** [comment]
Any plans for an EnhancedDynamoDbClient in Micronaut? https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/dynamodb-enhanced-client.html


We just do it in a library so the Beans are available for use. I think we had to do something custom with `@Secondary` constructors to make it work with micronaut test resources as well.

**Reply by sdelamo** [reply]
I want to have the best dynamo integration possible @scprek 

lets move this discussion to https://github.com/micronaut-projects/micronaut-aws/pull/1664

**Reply by scprek** [reply]
Ok moved there

**Comment by wmunyan** [comment]
Thank you for including more commentaries on the benefits of Micronaut in relation to Spring Boot and/or other frameworks!  Something I am continuously coming up against in my current role is trying to convince and influence in favor of Micronaut.  I'm always presenting the benefits of Micronaut, but am always told Spring Boot is "the way its always been done" and "that's the experience people have".  I'm definitely appreciative of any hard data I can use to support Micronaut!

**Comment by KangoV** [comment]
I'd love to see Micronaut Serialisation support the Builder pattern as Jackson does.
More guides on how to use Micronaut in a standalone setting would be very useful. For example, I'm creating a STIX parser library using Serialisation and Validation (instead of Jackson Databind and Hibernate Validator). Lots of roadblocks just getting serialisation to work.

**Reply by sdelamo** [reply]
@KangoV can you create an issue in the [Micronaut Serialization repository](https://github.com/micronaut-projects/micronaut-serialization/issues) describing the requested feature?

**Reply by yawkat** [reply]
The builder pattern should already be supported, we did a lot of work on that recently

**Comment by sdelamo** [comment]
I talked with the Micronaut Boards of directors and they suggested additional topics: [mybatis](https://github.com/micronaut-projects/micronaut-sql/issues/1177) and a [Cloud Gateway Solution](https://github.com/micronaut-projects/micronaut-core/issues/10164).

**Reply by jeffscottbrown** [reply]
> I talked with the Micronaut Boards of directors

To clarify, I don't think that was the board of directors.  I think that was the Technology Advisory Board.

**Reply by sdelamo** [reply]
Yes, it was the Technology Advisory Board.

**Comment by sdelamo** [comment]
# Micronaut 4.3.0 Plan Summary

- [4.3.0 Blog post announcement](https://micronaut.io/2024/02/06/micronaut-framework-4-3-0-released/)

## Advocacy

❌ **Youtube Channel**. Release ten videos of less than 15m for the YouTube channel covering features of the framework.
Guides

❌ Release 4 Guides comparing Micronaut Framework to Spring Boot. 

## New Modules

❌ Load Shedding
✅ Chatbots
❌ SourceGen

### Finish in progress Features
❌ AWS Better integration with DynamoDB.

## Enhancements

### Expression Language

✅ Add conditional caching with the new EL.
✅ Support Expression Language for Tenant resolution

### Others
❌ Integration with synk
✅ Eclipse Store
✅ Update to Hibernate 6.4.4.Final

### Security
✅ [Add nonreactive API for authentication fetcher](https://github.com/micronaut-projects/micronaut-security/issues/1476).
